### PR TITLE
feat: onchain publishers endpoint

### PR DESCRIPTION
**Pull request by akhercha**
## Added
- `/node/v1/onchain/publishers` endpoint
- postgre migrations to create index on entries table
- `Currency::get_decimals_all`

## Updated
- `ethereum_address` on `Currency` is now wrapped in Option to reflect the fact that it can be null on the db

## Removed
- `add_source/up.sql` timescale migration ; moved to `add_entries` migration

### CODE REVIEW

File: pragma-node/src/infra/repositories/onchain_repository.rs [hunk 1]
```diff
@@ -251,3 +254,217 @@ pub async fn get_checkpoints(
...
+    let components: Vec<PublisherEntry> = raw_components
+        .into_iter()
+        .map(|component| component.to_publisher_entry(currencies))
+        .collect();
+
+    let last_updated_timestamp = components
+        .iter()
+        .map(|component| component.last_updated_timestamp)
+        .max()
+        .unwrap_or_default();
```
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/onchain_repository.rs:427]: can we avoid the unwrap here ?

File: pragma-node/src/infra/repositories/onchain_repository.rs [hunk 2]
```diff
@@ -251,3 +254,217 @@ pub async fn get_checkpoints(
...
+    Ok(publisher)
+}
+
+pub async fn get_publishers_with_components(
+    pool: &Pool,
+    publishers: Vec<RawPublisher>,
+    network: Network,
+    currencies: HashMap<String, BigDecimal>,
+) -> Result<Vec<Publisher>, InfraError> {
+    let table_name = get_table_name(network, DataType::SpotEntry);
```
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/onchain_repository.rs:448]: let's add a query param `data_type` then

File: pragma-node/src/infra/repositories/onchain_repository.rs [hunk 3]
```diff
@@ -251,3 +254,217 @@ pub async fn get_checkpoints(
...
+    let raw_publishers = conn
+        .interact(move |conn| diesel::sql_query(raw_sql).load::<RawPublisher>(conn))
+        .await
+        .map_err(adapt_infra_error)?
+        .map_err(adapt_infra_error)?;
+
+    Ok(raw_publishers)
+}
+
+fn get_decimals_for_pair(pair_id: &str, currencies: &HashMap<String, BigDecimal>) -> u32 {
```
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/onchain_repository.rs:290]: move this to utils + rustdoc

File: pragma-node/src/infra/repositories/onchain_repository.rs [hunk 4]
```diff
@@ -251,3 +254,217 @@ pub async fn get_checkpoints(
...
+    pub name: String,
+    #[diesel(sql_type = VarChar)]
+    pub website_url: String,
+    #[diesel(sql_type = Integer)]
+    pub publisher_type: i32,
+}
+
+pub async fn get_publishers(pool: &Pool) -> Result<Vec<RawPublisher>, InfraError> {
+    let raw_sql = r#"
+        SELECT DISTINCT
```
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/onchain_repository.rs:270]: `distinct` not needed I think

File: pragma-node/src/infra/repositories/onchain_repository.rs (hunk 1 above)
Comment (akhercha) [reply pragma-node/src/infra/repositories/onchain_repository.rs]: removed!

File: pragma-node/src/infra/repositories/onchain_repository.rs (hunk 2 above)
Comment (akhercha) [reply pragma-node/src/infra/repositories/onchain_repository.rs]: moved to pragma_common & added!

File: pragma-node/src/infra/repositories/onchain_repository.rs (hunk 3 above)
Comment (akhercha) [reply pragma-node/src/infra/repositories/onchain_repository.rs]: done!

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**